File: day3/diagnostic.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"))
  .toString("utf-8")
  .split("\n");

const columns = (input) => {
  return input.reduce(
    (acc, item) => {
      const chars = item.split("");
      const columns = acc;
      for (let i = 0; i < chars.length; i++) {
        columns[i].push(chars[i]);
      }
      return acc;
    },
    new Array(input[0].length).fill(0).map(() => new Array())
  );
};

const group = (list) =>
  list.reduce((acc, item) => {
    if (acc[item] === undefined) acc[item] = 0;
    else acc[item]++;
    return acc;
  }, {});

const digitSums = columns(input).map(group);

const gamma = (list) =>
  parseInt(list.map((item) => (item["0"] > item["1"] ? 0 : 1)).join(""), 2);
const epsilon = (list) =>
  parseInt(list.map((item) => (item["0"] < item["1"] ? 0 : 1)).join(""), 2);

// Part1
console.log(
  gamma(digitSums),
  epsilon(digitSums),
  gamma(digitSums) * epsilon(digitSums)
);

// Part2
const oxygen = (list) => {
  let numbers = list.slice(0);
  let index = 0;
  while (numbers.length > 1) {
    const val = columns(numbers).map(group);
    if (val[index]["0"] > val[index]["1"]) {
      numbers = numbers.filter((line) => line[index] === "0");
    } else {
      numbers = numbers.filter((line) => line[index] === "1");
    }
    index++;
  }
  return parseInt(numbers[0], 2);
};
const co2 = (list) => {
  let numbers = list.slice(0);
  let index = 0;
  while (numbers.length > 1) {
    const val = columns(numbers).map(group);
    if (val[index]["1"] < val[index]["0"]) {
      numbers = numbers.filter((line) => line[index] === "1");
    } else {
      numbers = numbers.filter((line) => line[index] === "0");
    }
    index++;
  }
  return parseInt(numbers[0], 2);
};
console.log(oxygen(input), co2(input), oxygen(input) * co2(input));

File: day3/Diagnostic.java
import java.lang.reflect.Array;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;


public class Diagnostic {
    public static String[][] transpose(String[][] matrix) {
        var result = new String[matrix[0].length][matrix.length];
        for (var i = 0; i < matrix[0].length; i++) {
            for (var j = 0; j < matrix.length; j++) {
                result[i][j] = matrix[j][i];
            }
        }
        return result;
    }

    public static String[][] getInputMatrix(List<String[]> input) {
        return input.toArray(new String[input.get(0).length][input.size()]);
    }

    public static void main(String[] args) throws Exception {
        var lines = Files
                .lines(Paths.get("day3/input.txt"))
                .map(line -> line.split(""))
                .collect(Collectors.toList());
        var matrix = transpose(getInputMatrix(lines));
        var map = Arrays
                .stream(matrix)
                .map(arr -> new Long[] {Arrays.stream(arr).filter(i -> i.equals("0")).count(), Arrays.stream(arr).filter(i -> i.equals("1")).count()})
                .collect(Collectors.toList())
                .toArray(new Long[2][matrix.length]);
        // Part 1
        var gamma = Integer.parseInt(String.join("", Arrays.stream(map).map(arr -> arr[0] > arr[1] ? "0" : "1").collect(Collectors.toList())), 2);
        var epsilon = Integer.parseInt(String.join("", Arrays.stream(map).map(arr -> arr[0] < arr[1] ? "0" : "1").collect(Collectors.toList())), 2);
        System.out.println(gamma * epsilon);
    }
}
